src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+
+	creds, err := api.prepareTLSToAgent(log)
+	if err != nil {
+		return err
+	}
+
+	req, err := stream.Recv()
+	if err != nil {
+		return errorf(codes.Unimplemented, "unable to receive message: %w", err)
+	}
+
+	opts, isStart := req.Operation.(*CaptureRequest_Start)
+	if !isStart {
+		return fmt.Errorf("expected start message, got %v: %w", req.Operation, errUnexpectedMessage)
+	}
+
+	targets, resolveErr := api.resolveAgentEndpoints(opts.Start.Capture, log)
+	if errors.Is(resolveErr, errValidationFailed) {
+		return errorf(codes.InvalidArgument, "capture targets not found: %w", err)
+	}
+
+	streamPreparer := &streamPrep{}
+
+	// Start capture
+	out, err := api.capture(ctx, stream, streamPreparer, opts.Start.Options, targets, creds, log)
+	if err != nil {
+		return err
+	}
+
+	forwardWG := &sync.WaitGroup{}
+	forwardWG.Add(1)
+
+	forwardToStream(cancel, out, stream, api.bufConf, forwardWG, api.id)
+
+	// Wait for capture stop
+	stopCmd(cancel, stream)
+
+	err = Cause(ctx)
+	// Cancelling the context with nil causes context.Cancelled to be set
+	// which is a non-error in our case.
+	if err != nil {
+		return err
+	}
+
+	forwardWG.Wait()
+
+	api.deregisterStream(ctx, &stream)
+
+	return nil
+}
+
+func (api *API) prepareTLSToAgent(log *zap.Logger) (credentials.TransportCredentials, error) {
+	if api.agents.MTLS == nil || api.agents.MTLS.SkipVerify {
+		return insecure.NewCredentials(), nil
+	}
+
+	// Load certificate of the CA who signed agent's certificate
+	pemAgentCA, readErr := os.ReadFile(api.agents.MTLS.CertificateAuthority)
+	if readErr != nil {
+		log.Error("Load Agent CA certificate failed")
+		return nil, readErr
+	}
+
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(pemAgentCA) {
+		return nil, fmt.Errorf("failed to add agent CA's certificate")
+	}
+
+	// Create the credentials and return it
+	config := &tls.Config{
+		RootCAs:    certPool,
+		ServerName: api.agents.MTLS.CommonName,
+		MinVersion: tls.VersionTLS12,
+		MaxVersion: tls.VersionTLS13,
+	}
+
+	// Load client's certificate and private key
+	if api.agents.MTLS.Certificate != "" && api.agents.MTLS.PrivateKey != "" {
+		clientCert, err := tls.LoadX509KeyPair(api.agents.MTLS.Certificate, api.agents.MTLS.PrivateKey)
+		if err != nil {
+			log.Error("Load API client certificate or private key failed")
+			return nil, err
+		}
+		config.Certificates = []tls.Certificate{clientCert}
+	}
+
+	return credentials.NewTLS(config), nil
+}
+
+// resolveAgentEndpoints tries all registered api.handlers until one responds or none can be found that
+// support this capture request. The responsible handler is then queried for the applicable pcap-agent endpoints corresponding to this capture request.
+func (api *API) resolveAgentEndpoints(capture *Capture, log *zap.Logger) ([]AgentEndpoint, error) {
+	for name, handler := range api.handlers {
+		if handler.canHandle(capture) {
+			log.Sugar().Debugf("Resolving agent endpoints via handler %s for capture %s", name, capture)
```

Comment:
Why? Creating a log parser that handles a few known attributes is fine, adding an arbitrary amount of fields could be counterproductive.

Reply:
Isn't the whole idea of structured logging to not having to write a parser? It spits out `json` (or whatever your preferred format is) and your logging system already knows how to pick up on that.

The main reason I pointed this one out is that is one of two occurrences and I was wondering why exactly at those two positions we would want to use it but nowhere else.

Reply:
Yes that's fair. I wanted to avoid having 'indefinite' amounts of fields, but you're right. The parser will be "read json please" instead of anything else. Not all of those fields would need to be indexed for faster search.

Reply:
As discussed, we do not want to use sugared logger at this place.